### ISSUE TITLE
Crash on LAPACK dsyev and INTERFACE64=0

### ISSUE THREAD

**Issue by perecn** [xianyi/OpenBLAS#1355]
When calling the Lapack's eigensolver dsyev_ from a C++ program I found that OpenBlas consistently crashes. This happens on OpenSuse 42.3 (latest) x86_64, haswell. On the debugger, trace goes from dsyev_ to dgemm. I found this on versions 0.2.18 to the latest on git. I compiled (gcc, gfortran) OpenBlas using both

make USE_THREAD=0 USE_OPENMP=0 INTERFACE64=1
make USE_THREAD=0 USE_OPENMP=0 INTERFACE64=0 (as is in OpenSuse rpm)

combined with all possible declarations
extern "C"{ ....
void dsyev_(const char* jobz, const char* uplo, int* n,  ...
void dsyev_(const char* jobz, const char* uplo, long* n,  ...
void dsyev_(const char* jobz, const char* uplo, unsigned long* n,  ...
}

and a matrix size of N=1000 or 2000 (for very tiny matrices ~50 there is no crash)

Lapack's dsyev only works when OpenBlas is compiled using 8 bit ints (INTERFACE64=1) and using long in declaration. In all other cases it crashes. Note that I use 4 bit ints in my program, OpenSuse default is 4 too.

My guess is that it could be some int long miss-coding somewhere in the OpenBlas internals, from dsyev to dgemm call line.

Thank you for your attention.

**Comment by martin-frbg** [comment]
Would it be possible for you to provide a short demo program that reproduces this error, or at least provide a debugger trace from an OpenBLAS built with DEBUG=1 so that we get the exact line where it crashes ? Running your program under valgrind (again ideally with a debuggable build of OpenBLAS) is also likely to shed a light on your problem. 
Offhand  I am not aware of any such issue having been reported in the past, although INTERFACE64=0 is the default. I suspect in your use case it might make sense to call the LAPACKE C binding for dsyev instead of the FORTRAN function to avoid conflicts between the different calling conventions and array ordering semantics of the two programming languages.

**Comment by perecn** [comment]
Thank you. It does work, there is no bug. I compiled OpenBlas with DEBUG=1 on a minimal program and it worked fine. Then I rearranged the lib list in my real program, putting OpenBlas first and the crash went out. I found that one of the libs, not actually called in my test, has either a dgemm_ definition or a stand alone dgemm.

Thank you Martin also for your suggestion regarding LAPACKE. I checked the code time ago and I found too many copies and allocations to my taste. Instead I use a set of matrix classes already arranged with column major layout and calling BLAS/LAPACK and others is very smooth.

Best regards.

--------------------------------------------------------------------------------------
==12392== Use of uninitialised value of size 8
==12392==    at 0x968C6A: dgemm_
==12392==    by 0x9E9377: dlarfb_ (dlarfb.f:388)
==12392==    by 0x9E6A18: dorgql_ (dorgql.f:273)
==12392==    by 0x9D3F15: dorgtr_ (dorgtr.f:222)
==12392==    by 0x97DC0E: dsyev_ (dsyev.f:262)
==12392==    by 0x753EAE: mycheck (check.cpp:602)
 ...
==12392==  Uninitialised value was created by a stack allocation

 ...
==12392== Process terminating with default action of signal 11 (SIGSEGV)
==12392==  Access not within mapped region at address 0x19557000
==12392==    at 0x968C6F: dgemm_
--------------------------------------------------------------------------------------
#include <iostream>
extern "C"
{
    void dsyev_(const char* jobz, const char* uplo, int* n, double* a, int* lda, double* w, double* work, int* lwork, int* info);
}
int main()
{
    const char jobz = 'V';          // Eigenvalues and eigenvectors
    const char uplo = 'U';          // Upper triangular
    int n = 150;
    int info;
    double* eigenvalues = new double[n];
    double* eigenvectors = new double[n * n];
    for (int i = 0; i < n * n; ++i)
        eigenvectors[i] = 0.1;
    int lwork = 10 + (64 + 2) * n;
    double* work = new double[lwork];
    dsyev_(&jobz, &uplo, &n, eigenvectors, &n, eigenvalues, work, &lwork, &info);

    std::cout << n << std::endl;
    std::cout << info << std::endl;
    for (int i = 0; i < n; ++i)
        std::cout << i << ' ' << eigenvalues[i] << std::endl;

    delete [] eigenvectors;
    delete [] eigenvalues;
    delete [] work;
    return 0;
}

**Comment by brada4** [comment]
Do you still have full log? Can you dig back what (and by which function) was allocated right below 0x19557000 ?

**Comment by perecn** [comment]
I am sorry, I do not have the log anymore. What I posted were 2 pieces from the log. I cut first piece right after function 'mycheck' in the whole, failing program. The second piece was more or less the same. In any case, I was linking the whole program, not the minimalist I attached, with a solver that has a fake dgemm_, declared with 8 byte ints (long), and this interfered with OpenBlas in my whole program. There is no issue at all with OpenBlas, it works wonderful.
